src/google/protobuf/descriptor.cc

```diff
@@ -7875,23 +7875,36 @@ void DescriptorBuilder::ValidateFieldFeatures(
   }
 
   // Validate explicitly specified features on the field proto.
-  if ((field->containing_oneof() != nullptr || field->is_repeated() ||
-       field->message_type() != nullptr) &&
-      field->proto_features_->field_presence() == FeatureSet::IMPLICIT) {
-    AddError(
-        field->full_name(), proto, DescriptorPool::ErrorCollector::NAME,
-        "Only singular scalar fields can specify implicit field presence.");
-  }
-  if ((field->containing_oneof() != nullptr || field->is_repeated()) &&
-      field->proto_features_->field_presence() == FeatureSet::LEGACY_REQUIRED) {
-    AddError(
-        field->full_name(), proto, DescriptorPool::ErrorCollector::NAME,
-        "Only singular scalar fields can specify required field presence.");
+  if (field->proto_features_->has_field_presence()) {
+    if (field->containing_oneof() != nullptr) {
+      AddError(field->full_name(), proto, DescriptorPool::ErrorCollector::NAME,
+               "Oneof fields can't specify field presence.");
+    } else if (field->is_repeated()) {
+      AddError(field->full_name(), proto, DescriptorPool::ErrorCollector::NAME,
+               "Repeated fields can't specify field presence.");
+    } else if (field->is_extension() &&
+               field->proto_features_->field_presence() !=
+                   FeatureSet::LEGACY_REQUIRED) {
+      // Note: required extensions will fail elsewhere, so we skip reporting a
+      // second error here.
+      AddError(field->full_name(), proto, DescriptorPool::ErrorCollector::NAME,
+               "Extensions can't specify field presence.");
+    } else if (field->message_type() != nullptr &&
+               field->proto_features_->field_presence() ==
+                   FeatureSet::IMPLICIT) {
+      AddError(field->full_name(), proto, DescriptorPool::ErrorCollector::NAME,
+               "Message fields can't specify implicit presence.");
+    }
   }
   if (!field->is_repeated() &&
       field->proto_features_->has_repeated_field_encoding()) {
     AddError(field->full_name(), proto, DescriptorPool::ErrorCollector::NAME,
-             "Only repeated fields can specify `repeated_field_encoding`.");
+             "Only repeated fields can specify repeated field encoding.");
+  }
+  if (field->cpp_type() != FieldDescriptor::CPPTYPE_MESSAGE &&
+      field->proto_features_->has_message_encoding()) {
+    AddError(field->full_name(), proto, DescriptorPool::ErrorCollector::NAME,
+             "Only message fields can specify message encoding.");
   }
```

Comment:
@mkruskal-google, thanks for tagging me in this fix! BTW, both of these checks also apply to map types, which raises a couple of questions:
* Should a map type be allowed to indicate repeated field encoding? It would be a no-op for repeated field encoding since the repeated value is not a numeric scalar.
* Should a map type be allowed to indicate message encoding, and if so, does the encoding apply to just a possible message value or also to how the map entries themselves are encoded?

Reply:
nice, I'm really glad that worked, I was trying to figure out a way to mark you as a reviewer! :P

Those are good questions about maps.  I think the answer is that we should be banning both of these for different reasons:
- We should be validating that the repeated field is packable (similar to what we do for packed), so I'll add a check for that which should cover maps.
- Message encoding could in theory apply to maps, but I'd rather not support that new behavior without any demand for it (especially since the semantics are confusing).  IIUC today groups wouldn't be usable in maps.  You could reuse the generated message in a map, but it would just be encoded as a regular sub-message